### ISSUE TITLE
[css-contain] Should `contentvisibilityautostatechange` events bubble?

### ISSUE THREAD

**Issue by knowler** [w3c/csswg-drafts#11310]
This is more-so related to the associated event for `content-visibility: auto` (i.e. `contentvisibilityautostatechange`). I noticed that in Chrome (131) this bubbles, while in Safari (18.1) and Firefox (133) it does not. I was planning on making a WPT for this, but I don’t really know what the behaviour should be, so I figured I’d ask here.

I made [a demo highlighting the bubbling problem](https://knowler.dev/demos/WA2d_f9).

**Comment by knowler** [comment]
I was discussing this with @AmeliaBR on Mastodon and [she pointed out](https://front-end.social/@AmeliaBR/113584190196184251) that the parent event likely specifies this (i.e. in this case `Event` and its init `EventInit` interface). It seems like if bubbling was intended, then it was missed in the spec.

**Comment by knowler** [comment]
There is a Chrome issue open for this: https://issues.chromium.org/issues/365168180

I still think this is a spec issue where the designers of the feature forgot to set that it should bubble.

**Comment by LeaVerou** [comment]
This seems like a no brainer to me but we still need a resolution. Hopefully an async one will do.

**Comment by astearns** [comment]
The CSSWG will automatically accept this resolution one week from now if no objections are raised here. Anyone can add an emoji to this comment to express support. If you do not support this resolution, please add a new comment.

Proposed Resolution: The `contentvisibilityautostatechange` event should bubble

**Comment by smaug----** [comment]
To be consistent with things like toggle event, it should not bubble. That way nestedly dispatched events don't disturb event listeners in ancestors. And one can always use capturing listener, if one explicitly needs to handle nested stuff

**Comment by knowler** [comment]
I believe @vmpstr authored the original explainer where it was intended to bubble. Was there a good reason for that?

Original explainer: https://github.com/vmpstr/web-proposals/blob/main/explainers/cv-auto-event.md#proposal-contentvisibilityautostatechange

I don’t have a strong opinion (other than getting this fixed). Existing code that currently relies on the event needs to use a capturing listener to work across browsers.

**Comment by LeaVerou** [comment]
I think the rule of thumb is _would bubbling prevent more bugs than it would cause?_

I don't have a strong opinion either way. The fact that it was originally envisioned to bubble makes me lean towards bubbling. But beyond that, I agree with @smaug---- .

**Comment by vmpstr** [comment]
I admit that I didn't have any use cases that specifically require bubbling. Blink currently implements this as bubbling, so it would be a compat risk to change that in the implementation. As for whether the spec should change or not, I don't really have a strong opinion.

Edit:
I do remember what case it was: we had requests to basically do things like pausing video (and stopping other work) when content visibility auto skips its content and resuming video (and resuming other work) when it becomes rendered, and so in those cases it might be reasonable to use this behavior in an ancestor element and observe a whole set of elements. In which case we want bubbling. In fact, it's probably how a developer would implement that, since it's not tied to a specific element a lot of time (like a virtual scroller).

**Comment by vmpstr** [comment]
@smaug---- just a question out of curiosity, is toggle event not bubbling per spec because none of the events are bubbling unless explicitly stated that "bubbles is initialized to true". I just can't find any reference to bubbling in https://html.spec.whatwg.org/multipage/interaction.html#the-toggleevent-interface

**Comment by smaug----** [comment]
https://dom.spec.whatwg.org/#concept-event-fire:~:text=Fire%20in%20the%20context%20of%20DOM%20is%20short%20for%20creating%2C%20initializing%2C%20and%20dispatching%20an%20event%2E%20Fire%20an%20event%20makes%20that%20process%20easier%20to%20write%20down%2E

And that use case... just use a capturing event listener in such perhaps a bit unusual case.

**Comment by vmpstr** [comment]
Thank you for the spec reference, although I still fail to see any description of bubbles being false, other than in IDL function signatures. The rest of the spec references things like "set `bubbles` to _bubbles_" but doesn't really specify what is the value of _bubbles_ if it isn't explicitly stated. Anyway, that's a bit besides the point, it was just a curiosity.

> And that use case... just use a capturing event listener in such perhaps a bit unusual case.

I disagree that the case here is unusual. This is the intended use of the `contentvisibilityautostatechange` event, as outlined in the explainer. Toggling work for some subtree is frequently a more involved process than adding `content-visibility: auto` to some children of an element, so forcing the developer to necessarily add an event listener to each element affected by `content-visibility: auto` is awkward. Hence the case of observing such changes in some ancestor, I would argue, is the common use case for the event.

I would also push back on just using a capturing event listener. Yes, that can work, but bubbling the event also works. I think bubbling is a bit more flexible in that it allows multiple levels of ancestors to react in an appropriate way to the event. 

Is there a positive argument for not bubbling?

**Comment by knowler** [comment]
I agree that the use case for bubbling is not unusual (e.g. a long list of items that have content skipped for rendering) and I’d almost argue that it’s one of the more popular ones, as least as far as demonstrating the feature go.

**Comment by smaug----** [comment]
> I still fail to see any description of bubbles being false

Does [this](https://dom.spec.whatwg.org/#concept-event-create:~:text=Let%20dictionary%20be%20the%20result%20of%20converting%20the%20JavaScript%20value%20undefined%20to%20the%20dictionary%20type%20accepted%20by%20eventInterface%E2%80%99s%20constructor%2E) help?

**Comment by smaug----** [comment]
> Is there a positive argument for not bubbling?

The fact that if you add a bubbling-phase listener to your content-visibility element, you don't need to check event.eventPhase == Event.AT_TARGET in order to ensure the event is really for your element.
This shows up even more with oncontentvisibilityautostatechange event handler.
(Hmm, https://github.com/whatwg/html/pull/10364 is still open )

**Comment by vmpstr** [comment]
Consider a case with a scroller component that manages a bunch of `content-visibility: auto` elements. Also consider that one of the elements within this scroller is a nested scroller component that also manages a bunch of `content-visibility: auto` elements. If each of the scrollers wants to handle their own rendering and listen to this event, then I believe that a capturing event listener is quite difficult to manage (specifically because you want the innermost scroller to handle the event). Whereas doing `stopPropagation()` or setting it as default handled in the inner scroller so the outer doesn't handle it seems easier. Am I mistaken?

If I understand correctly event.eventPhase == Event.AT_TARGET is already a very common way to do the handling on the element itself.

I'd like to understand from more philosophical perspective why should we _prefer_ non bubbling vs bubbling? As I understand it we shouldn't bubble if the only sensible thing to do with an event is to handle it at target only. This isn't one of these events. Are there other  principles that cause us to prefer non bubbling?

**Comment by LeaVerou** [comment]
> This isn't one of these events. Are there other principles that cause us to prefer non bubbling?

I suggested one earlier: 
> _would bubbling prevent more bugs than it would cause?_

bugs that it would cause = unintentionally listening to events from descendants when you didn't intend to (think of the old `mouseover`/`mouseout` bugs which led to `mouseenter`/`mouseleave`)

I think a better formulation might be are there more cases that need to listen to descendants or more cases where that would cause bugs? 

`capture: true` is essentially an explicit opt-in to event delegation. I can't think of any downside to it except that it's less common. So I think even 50-50 isn't enough, there would need to be _significantly_ more cases that need delegation than cases where that's a bug.
